smartsim/entity/_new_ensemble.py

```diff
@@ -0,0 +1,109 @@
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import copy
+import itertools
+import os
+import typing as t
+
+from smartsim.entity import _mock, entity, strategies
+from smartsim.entity.files import EntityFiles
+from smartsim.entity.model import Application
+from smartsim.entity.strategies import ParamSet
+from smartsim.launchable.job import Job
+
+if t.TYPE_CHECKING:
+    from smartsim.settings.launchSettings import LaunchSettings
+
+
+# TODO: If we like this design, we need to:
+#         1) Move this to the `smartsim._core.entity.ensemble` module
+#         2) Decide what to do witht the original `Ensemble` impl
+class Ensemble(entity.CompoundEntity):
+    """Entity to help parameterize the creation multiple application
+    instances.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        exe: str | os.PathLike[str],
+        exe_args: t.Sequence[str] | None = None,
+        exe_arg_parameters: t.Mapping[str, t.Sequence[t.Sequence[str]]] | None = None,
+        files: EntityFiles | None = None,
+        file_parameters: t.Mapping[str, t.Sequence[str]] | None = None,
+        permutation_strategy: str | strategies.PermutationStrategyType = "all_perm",
+        max_permutations: int = -1,
+        replicas: int = 1,
+    ) -> None:
+        self.name = name
+        self.exe = os.fspath(exe)
+        self.exe_args = list(exe_args) if exe_args else []
+        self.exe_arg_parameters = (
+            copy.deepcopy(exe_arg_parameters) if exe_arg_parameters else {}
+        )
+        self.files = copy.deepcopy(files) if files else EntityFiles()
+        self.file_parameters = dict(file_parameters) if file_parameters else {}
```

Comment:
do you mind adding a deep_copy for the file_parameters, just to be consistent with exe_arg_parameters? That is a my bad MERP

Reply:
What are your thoughts on this, I am thinking that the attributes that do not need to be deepcopied are replicas, max_permutations and strategy, just bc they do not live on to Application and are not reused

Reply:
or hmm maybe deepcopy is just something to worry about for Application?

Reply:
I think the general thought process is that we should make copies of anything that is mutable when it is past to the ensemble constructor.

Because strings are immutable in python, a deep-copy of a `dict[str, str]` (or really any `MutableMapping[str, str]`) is functionally equivalent to making a shallow copy! I chose `dict` over `copy.copy` or `copy.deepcopy` simply so that any helper methods know the actual type of the container, and so that user know they could modify the collection after construction (should they choose)
```py
original_params = {"SOME": "PARAM"}
ens = Ensemeble(..., file_parameters=original_params, ...)

# Wait I need to change those params for some reason!!
# But know that I can because the attr is typed as a dict!
ens.file_parameters["SOME"] = "OTHER_PARAMETER"  # passes type check!
assert ens.file_parameters != original_params    # passes!
```

Reply:
To that end, I think we need:
- a deep copy of `files` as its a complex user defined class
- a deep copy of `exe_args_parameters` as its a container of containers of immutable
- shallow copies of `file_parameters` and `exe_args` as they are containers of immutables
- No copies of any other parameters as they are all immuatable

What do you think?